db/schema_applier.cc

```diff
@@ -0,0 +1,930 @@
+static read_schema_for_keyspaces(distributed<service::storage_proxy>& proxy, const sstring& schema_table_name, const std::set<sstring>& keyspace_names)
+{
+    auto map = [&proxy, schema_table_name] (const sstring& keyspace_name) { return read_schema_partition_for_keyspace(proxy, schema_table_name, keyspace_name); };
+    auto insert = [] (schema_result&& result, auto&& schema_entity) {
+        if (!schema_entity.second->empty()) {
+            result.insert(std::move(schema_entity));
+        }
+        return std::move(result);
+    };
+    co_return co_await map_reduce(keyspace_names.begin(), keyspace_names.end(), map, schema_result{}, insert);
+}
+
+// Returns names of live table definitions of given keyspace
+future<std::vector<sstring>>
+static read_table_names_of_keyspace(distributed<service::storage_proxy>& proxy, const sstring& keyspace_name, schema_ptr schema_table) {
+    auto pkey = dht::decorate_key(*schema_table, partition_key::from_singular(*schema_table, keyspace_name));
+    auto&& rs = co_await db::system_keyspace::query(proxy.local().get_db(), schema_table->ks_name(), schema_table->cf_name(), pkey);
+    co_return boost::copy_range<std::vector<sstring>>(rs->rows() | boost::adaptors::transformed([schema_table] (const query::result_set_row& row) {
+        const sstring name = schema_table->clustering_key_columns().begin()->name_as_text();
+        return row.get_nonnull<sstring>(name);
+    }));
+}
+
+// Applies deletion of the "version" column to system_schema.scylla_tables mutation rows
+// which weren't committed by group 0.
+static void maybe_delete_schema_version(mutation& m) {
+    if (m.column_family_id() != scylla_tables()->id()) {
+        return;
+    }
+    const column_definition& origin_col = *m.schema()->get_column_definition(to_bytes("committed_by_group0"));
+    const column_definition& version_col = *m.schema()->get_column_definition(to_bytes("version"));
+    for (auto&& row : m.partition().clustered_rows()) {
+        auto&& cells = row.row().cells();
+        if (auto&& origin_cell = cells.find_cell(origin_col.id); origin_cell) {
+            auto&& ac = origin_cell->as_atomic_cell(origin_col);
+            if (ac.is_live()) {
+                auto dv = origin_col.type->deserialize(managed_bytes_view(ac.value()));
+                auto committed_by_group0 = value_cast<bool>(dv);
+                if (committed_by_group0) {
+                    // Don't delete "version" for this entry.
+                    continue;
+                }
+            }
+        }
+        auto&& cell = cells.find_cell(version_col.id);
+        api::timestamp_type t = api::new_timestamp();
+        if (cell) {
+            t = std::max(t, cell->as_atomic_cell(version_col).timestamp());
+        }
+        cells.apply(version_col, atomic_cell::make_dead(t, gc_clock::now()));
+    }
+}
+
+future<std::set<sstring>> merge_keyspaces(distributed<service::storage_proxy>& proxy, const schema_result& before, const schema_result& after)
```

Comment:
This actually looks like it could be private to the module (static)

Reply:
yes